### ISSUE TITLE
VM fails to boot when passing PCI SAS HBA through

### ISSUE THREAD

**Issue by msgerbs** [xcp-ng/xcp#213]
I am attempting to use PCI passthrough with a SAS HBA (specifically, an IBM M1015 crossflashed to 9211-8i IT mode). I followed [this guide ](https://github.com/xcp-ng/xcp/wiki/PCI-Passtrough) exactly and triple checked both the grub config and the VM config. When I try to start the VM, I get this error message: 

`The server failed to handle your request, due to an internal error.  The given message may give details useful for debugging the problem.
message: xenopsd internal error: (Failure
  "Error from xenguesthelper: xenguest: xc_dom_build_image: [1] panic: xc_dom_boot.c:178: xc_dom_boot_domU_map: failed ")
`

If I remove the PCI portion of the VM config, it boots without a problem. I verified that the PCI address shows up in `xl pci-assignable-list`. I tried rebooting the host. I verified that vt-d and all other virtualization options are enabled in the host BIOS. What else can I try?

Here is the relevant grub.cfg section:
`menuentry 'XCP-ng (Serial)' {
        search --label --set root root-kxaoyb
        multiboot2 /boot/xen.gz com1=115200,8n1 console=com1,vga dom0_mem=2048M,max:2048M watchdog ucode=scan dom0_max_vcpus=1-8 crashkernel=192M,below=4G
        module2 /boot/vmlinuz-4.4-xen root=LABEL=root-kxaoyb ro nolvm hpet=disable rd.auto console=tty0 xencons=hvc console=hvc0 xen-pciback.hide=(08:00.0)
        module2 /boot/initrd-4.4-xen.img
}
`
And here is the relevant section of vm config:
`                          other-config (MRW): pci: 0/0000:08:00.0; base_template_name: Other install media; import_task: OpaqueRef:7e4d5655-34d6-41ff-a7de-d5171f69d3f1; mac_seed: e40adbb8-cbe4-0ea1-59cd-5e27974c6a93; install-methods: cdrom`

**Comment by borzel** [comment]
Just to be sure: I assume your XCP-ng OS disk is NOT connected via this SAS HBA?

**Comment by msgerbs** [comment]
Right. The OS disk is connected to the onboard SATA controller.

**Comment by borzel** [comment]
you have to use `pci=0/0000:08:00.0` not `pci: 0/0000:08:00.0`

**Comment by msgerbs** [comment]
This is the exact command that I ran: `xe vm-param-set other-config:pci=0/0000:08:00.0 uuid=4910a900-2deb-99f6-6b47-48e7c98592e5`

It just shows up with a colon when printing the configuration.

**Comment by borzel** [comment]
Ok. Just another idea (https://patchwork.kernel.org/patch/9719689/): have you set different max ram and ram?

**Comment by msgerbs** [comment]
It is set to a fixed 8GB. I also tried 4GB RAM, and tried removing all NICs from the VM as well as totally recreating the VM and ended up with the same issue.

**Comment by msgerbs** [comment]
I think this is the issue: https://support.hpe.com/hpsc/doc/public/display?docId=emr_na-c04781229&sp4ts.oid=5249566 I installed Proxmox on the same host and ran into issues with PCI passthrough under KVM as well, leading me to that article.

I am attempting to figure out how to download the RPMs for the tools needed to run this fix, since they don't seem to be in any of HPs repositories... Will update here once I've had a chance to try it.

**Comment by msgerbs** [comment]
That was it! Was not an XCP-NG issue at all. Sorry for wasting your time.

**Comment by borzel** [comment]
@msgerbs no problem :-)

**Comment by andrewreid** [comment]
@msgerbs, I've come across this thread trying to fix the same issue on a HP ML350p G8. I've tried to follow the link you've shared, but HPE seem to want to keep that for paying customers!

Can you share what you learned about fixing this? I've found a few crumbs around the place suggesting there's an issue, but not what the solution is.

**Comment by msgerbs** [comment]
@andrewreid Unfortunately I never did get this working, at least not that I can recall. I'm not sure what I did at this point, but I do know I never got it working. The page didn't require login when I was working on this so I assume it's been removed. 

I did find this just now which might help: https://forum.proxmox.com/threads/pcie-pass-trough-in-a-hpe-ml350p-to-a-freenas-vm.70242/ These steps look vaguely familiar so I assume I did similar, but maybe it will work for you. I'm curious if you get it working, as it seems to be a common issue with this generation of HP servers (mine was an ML310e Gen 8).

**Comment by msgerbs** [comment]
This post also contains considerable detail and may help: https://forums.unraid.net/topic/54414-solved-hp-gen8-gpu-passthrough/

**Comment by andrewreid** [comment]
We must have found that link to the Proxmox forum at the same time! Installing the HP scripting tools and excluding the physical slot with the XML… that did the trick. 

Thanks for the follow-up. Hopefully this paper trail helps the next punter with the same problem. The reading I did today suggests that it certainly is a quirk of HPE fear of that vintage.

**Comment by olivierlambert** [comment]
Wow that's good news @andrewreid 

If you want to detail the steps, it might help others

**Comment by andrewreid** [comment]
From my reading, it seems that this is a known problem peculiar to HP (and maybe other enterprise vendors) servers of a certain vintage. My HP ML350p G8 was certainly one of these machines. This Proxmox [forum post](https://forum.proxmox.com/threads/pcie-pass-trough-in-a-hpe-ml350p-to-a-freenas-vm.70242/) details a lot of the issues, and some useful information which I used to fix the problem for xcp-ng.

On your xcp-ng host, you'll need to install [HP scripting tools](http://downloads.linux.hpe.com/SDR/project/stk/). I found that a `baseurl` set to `http://downloads.linux.hpe.com/repo/stk/rhel/7.3Server/x86_64/current/` yielded a result, and I could install the scripting tools with `yum install hp-scripting-tools`.

Then identify the physical slot the card you're looking to pass through is occupying, using `lspci -vmm | grep -B5 PhySlot`. In my case, it was physical slot 4.

Download the HP ConRep XML `wget https://downloads.hpe.com/pub/softlib2/software1/pubsw-linux/p1472592088/v95853/conrep_rmrds.xml` and then create an exclusion file referencing your physical slot obtained above. I created `exclude.xml` and entered:

```
<Conrep>
  <Section name="RMRDS_Slot4" helptext=".">Endpoints_Excluded</Section>
</Conrep>
```

Then execute the ConRep tool: `conrep -l -x conrep_rmrds.xml -f exclude.dat`

Reboot the host and pour yourself a beer.
